### PR TITLE
Add find() support within our DFU traverser

### PR DESCRIPTION
**Pull request by dongahn**
This PR introduces `find()` support within our DFU traverser
and integrates it with `resource-query` for testing.

`find()` traverses the resource graph, evaluate each visiting
vertex and emit those resources whose states are matched
with the criteria passed in.

- To evaluate each resource vertex, we view
    each resource vertex as a set of predicates:
   ` { predicate1, predicate2, ..., predicateN }`.
   Example predicates are "is status down?" (`status=down`),
   "is scheduled state allocated?" (`sched-now=allocated`)
    or "is future scheduled state reserved?" (`sched-future=reserved`).

- Add and use the `expr_eval_target_vtx_t` adapter class
   on each visiting resource vertex so that we can evaluate
   each queried predicate on its current state without having
   to modify the resource vertex classes themselves.

- Because we allow the partial marking of a vertex,
   the state of an ancestor vertex can often determine
   the effective state of a child vertex.
   For instance, if the status of a compute node vertex
   is "down", the status of a core underneath it must also be
   "down" even if the  actually marked status is still "up".
    To support this case, we add and use the
    `vtx_predicates_override_t` class to aid in passing
    ancestor predicates down to the offspring resources
    throughout the traversal.

- `find()` takes in a base writer object and match criteria.
    Performs the DFU visit on the resource graph and conducts
    the following on certain visit events:
    1) On each preorder visit, it builds the predicates
    of the visiting vertex and passes it to  the next vertex visit
    (depth first visit on the dominant subsystem and up visit on
    auxiliary subsystems).
    2) On each postorder visit, it builds `expr_eval_vtx_target_t`
    based on the state of the visiting vertex *and* the predicates set
    passed down from the parent and evaluates the
    expression on this mixed state.

Integrate `find()` with `resource-query` and
adds a test using its `find expression` subcommand.

Add a fix to `resource-query` for `set|get-status`
subcommands to avoid a race between resource-query's
prompt outputs and the testing output by supporting
`-t` option within  `set|get-status`. 
Adjust `t3024-resource-status.t` with respect to that.

### CODE REVIEW

**Comment by dongahn** [comment]
PR #684 depends on this PR. 

This PR also depends on one commit from PR #667, it would be best to merge PR #667 first.

**Comment by dongahn** [comment]
CI isn't happy; I added a few changes. Once Trivis CI turns green, this should be ready for your reviews.

**Comment by dongahn** [comment]
Travis is green now!

**Comment by dongahn** [comment]
@Mergifyio rebase

**Comment by dongahn** [comment]
@Mergefy.io rebase

**Comment by dongahn** [comment]
@Mergifyio rebase

File: resource/traversers/dfu_impl.hpp [hunk 1]
```diff
@@ -205,6 +209,31 @@ class dfu_impl_t {
...
+     *  \param criteria  matching criteria expression string. Each
+     *                   individual criterion is expressed as a key-value
+     *                   pair, representing a predicate p(x) where key is
+     *                   is p and value is x.
+     *                   Currently supported expressions are
+     *                   "status={up|down}", "sched-now={allocated|free}",
+     *                   "sched-future={reserved|free}, or any combination
+     *                   of them separated with "and", "or", or a whitespace
+     *                   which is also interpreted as "and" logical
+     *                   operator of two expressions. Parentheses
```
Comment (milroy) [inline resource/traversers/dfu_impl.hpp:225]: It looks like this answers my previous question about the whitespace in `resource/evaluators/test/expr_eval_test01.cpp`.  Can you add a similar explanation there?

File: t/t3025-resource-find.t [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+	find status=up
+	quit
+EOF
+    ${query} -L ${grugs} -F rv1_nosched -S CA -P high -t up.out < cmds002 &&
+    cat up.out | grep -v "INFO" > up.json &&
+    remove_times up.json > up.filt.json &&
+    diff up.filt.json full.filt.json
+'
+
+test_expect_success 'mark down 2 cores and find status=down' '
```
Comment (milroy) [inline t/t3025-resource-find.t:44]:
```suggestion
test_expect_success 'mark down 3 cores and find status=down' '
```

File: resource/utilities/command.cpp [hunk 3]
```diff
@@ -478,9 +478,12 @@ int cmd_set_status (std::shared_ptr<resource_context_t> &ctx,
         ctx->db->metadata.by_path.find (vtx_path);
```
Comment (milroy) [inline resource/utilities/command.cpp:478]: It's great that you found this race condition!

**Comment by milroy** [comment]
This contribution is really great!

Most of my suggestions and comments are on the order of nits.  The main exception is related to parsing parentheses.  My manual tests mostly behave as expected, but I did find some strange results related to an unbalanced terminating parenthesis.  Using the tiny example cluster for example:

```bash
resource-query> set-status /tiny0/rack0/node0/socket0/core0 down
resource-query> find (status=down) and (sched-now=free)
      ---------------core0[1:x]
      ------------socket0[1:x]
      ---------node0[1:x]
      ------rack0[1:x]
      ---tiny0[1:x]
INFO: =============================
INFO: EXPRESSION="(status=down) and (sched-now=free)"
INFO: =============================
resource-query> find (status=down) and (sched-now=free))
      ---------------core0[1:x]
      ------------socket0[1:x]
      ---------node0[1:x]
      ------rack0[1:x]
      ---tiny0[1:x]
INFO: =============================
INFO: EXPRESSION="(status=down) and (sched-now=free))"
INFO: =============================
resource-query> find (status=down) and (sched-now=free)))
ERROR: find: invalid criteria: (status=down) and (sched-now=free))).
```

**Comment by dongahn** [comment]
> I did find some strange results related to an unbalanced terminating parenthesis. Using the tiny example cluster for example:

Thanks @milroy. I will go through all of your comments. But it feels like this one is some "one-off" bug somewhere. I will look!

**Comment by milroy** [comment]
> But it feels like this one is some "one-off" bug somewhere.

Yes, that's the feeling I get since my other tests are successful.

File: resource/evaluators/expr_eval_api.cpp [hunk 4]
```diff
@@ -0,0 +1,352 @@
+/*****************************************************************************\
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/evaluators/expr_eval_api.cpp:1]: A nest parenthesized expression should be evaluated before the outer expression is evaluated.

File: resource/evaluators/expr_eval_target.hpp [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+/*! Expression evaluation target base abstract class.
+ *  Define the interfaces that expr_eval_api_t uses to evaluate
+ *  and validate a leaf-level expression predicate: p(x).
+ *  These interfaces are defined as pure virtual methods.
+ *  Therefore, a derived class must override and implement them.
+ */
+class expr_eval_target_base_t {
+public:
+
+    /*! Validate if a predicate expression is valid.
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/evaluators/expr_eval_target.hpp:43]: Thanks. Good catch.

File: resource/evaluators/expr_eval_vtx_target.hpp [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    void initialize (const vtx_predicates_override_t &p,
+                     const std::shared_ptr<
+                         const f_resource_graph_t> g, vtx_t u);
+
+    /*! Is this object initialized?
+     */
+    bool is_initialized () const;
+
+private:
+    bool m_initialized{false};
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/evaluators/expr_eval_vtx_target.hpp:110]: I actually have seen no leading whitespace version of uniform initialization more often. (But then I have seen no leading whitespace version of function calls more often :-)

Since we don't have a coding standard for this and I think I have used no leading whitespace version previously. We probably should keep it this way for now.

File: resource/evaluators/test/expr_eval_test01.cpp [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
+ *  See also:  http://www.gnu.org/licenses/
+\*****************************************************************************/
+
+#if HAVE_CONFIG_H
+# include <config.h>
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/evaluators/test/expr_eval_test01.cpp:26]: Great catch!

File: resource/evaluators/test/expr_eval_test01.cpp [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+    if ( (rc = validate (p, x)) < 0)
+        return rc;
+    result = true;
+    return rc;
+}
+
+void build_simple_expr (std::vector<std::string> &expr_vector)
+{
+    expr_vector.push_back ("  status=up      ");
+    expr_vector.push_back ("  status=up          sched-now=allocated    	");
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/evaluators/test/expr_eval_test01.cpp:72]: Yes this is intended. I wanted to test extra whitespace cases and also no operator between predicates that are assumed to be "and". (modeled from Unix `find` tool).

File: resource/evaluators/expr_eval_api.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/evaluators/expr_eval_api.cpp:1]: Better! Thanks.

File: resource/traversers/dfu_impl.cpp [hunk 9]
```diff
@@ -575,6 +575,71 @@ int dfu_impl_t::dom_dfv (const jobmeta_t &meta, vtx_t u,
...
+    } else if ( (rc = w->emit_vtx (level (), *m_graph, u,
+                                   (*m_graph)[u].size, true)) < 0) {
+        m_err_msg += __FUNCTION__;
+        m_err_msg += std::string (": error from emit_vtx: ") + strerror (errno);
+        goto done;
+    }
+    rc = nchildren + 1;
+done:
+    m_trav_level--;
+    return rc;
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/traversers/dfu_impl.cpp:640]: The higher number of subtree matches, the higher rc will be. It is just indicate whether you need to select the visiting vertex event if it doesn't actually match with the expression. To form a graph, you need to select that vertex in that case. It is not necessarily more desirable than `rc == 1`.

If you are asking why we don't limit `rc` to be 0, -1 or 1, I don't have a good answer. `rc > 1` seems more natural to propagate richer information to the upper function frame in this recursion. So I chose it this way.

File: resource/traversers/dfu_impl.hpp (hunk 1 above)
Comment (dongahn) [reply resource/traversers/dfu_impl.hpp:225]: Sure. Will do.

File: t/t3025-resource-find.t (hunk 2 above)
Comment (dongahn) [reply t/t3025-resource-find.t:44]: Oops. Thanks.

File: resource/evaluators/test/expr_eval_test01.cpp [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+    expr_vector.push_back ("sched-now=allocated or status=up");
+    expr_vector.push_back ("sched-now=free or status=up");
+    expr_vector.push_back ("sched-now=free or   status=up");
+    expr_vector.push_back ("sched-now=free or status=up");
+    expr_vector.push_back ("sched-now=allocated or status=down");
+    expr_vector.push_back ("sched-now=allocated and status=down");
+    expr_vector.push_back ("sched-now=allocated or status=down");
+    expr_vector.push_back ("sched-now=free status=down");
+    expr_vector.push_back ("sched-now=free and status=down");
+    expr_vector.push_back ("sched-now=free  or          status=down");
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/evaluators/test/expr_eval_test01.cpp:95]: This is intended: test to see if extra whitespaces are ignored.

**Comment by dongahn** [comment]
@Mergifyio rebase

**Comment by dongahn** [comment]
@milroy: Thank you for your review agains. I addressed all of your comments and added them as fixup commits. The only change that has already been squashed is for your commit message change request for the predicate. 

> But it feels like this one is some "one-off" bug somewhere. I will look!

As suspected, this was indeed due to an one-off bug where I should have use `<=` instead of `<`. 

Great finding!

File: t/t3025-resource-find.t [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+	match allocate ${node1_job}
+	match allocate ${sock1_job}
+	set-status /tiny0/rack0/node1 down
+	set-status /tiny0/rack0/node0/socket1 down
+	match allocate ${sock1_job}
+	match allocate_orelse_reserve ${sock1_job}
+	set-status /tiny0/rack0/node0/socket1 up
+	set-status /tiny0/rack0/node0/socket1/core19 down
+	set-status /tiny0/rack0/node0/socket0/core16 down
+	find (status=up and sched-now=allocated) or sched-future=reserved
```
Comment (SteVwonder) [inline t/t3025-resource-find.t:223]: Just to clarify my own understanding, is it possible for a node to be down and still have a new job reservation made on it?

File: resource/evaluators/expr_eval_api.cpp [hunk 12]
```diff
@@ -289,7 +289,7 @@ int expr_eval_api_t::validate (const std::string &e,
         goto done;
     at = next;
 
-    while (at < e.find_last_not_of (" \t")) {
+    while (at <= e.find_last_not_of (" \t")) {
```
Comment (SteVwonder) [inline resource/evaluators/expr_eval_api.cpp:292]: I didn't see the original comment from Dan that lead to this fixup commit, but is there a test (either unit or integration) that covers this edge case now?  Not sure how easy this would be to test for.

File: t/t3025-resource-find.t (hunk 11 above)
Comment (dongahn) [reply t/t3025-resource-find.t:223]: Yes it is possible under `resource-query` and our resource graph infrastructure. But how we use this in `sched-fluxion-resource`, this will not be possible as "down" marking will call sched_loop which will clear up all reservations and then rebuild. While rebuilding the alloc/reservations, the down resources won't be used.

File: t/t3025-resource-find.t [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+
+test_expect_success 'find status=up returns the entire system' '
+    cat > cmds002 <<-EOF &&
+	find status=up
+	quit
+EOF
+    ${query} -L ${grugs} -F rv1_nosched -S CA -P high -t up.out < cmds002 &&
+    cat up.out | grep -v "INFO" > up.json &&
+    remove_times up.json > up.filt.json &&
+    diff up.filt.json full.filt.json
```
[the remark this answers is not in the document]
Comment (dongahn) [reply t/t3025-resource-find.t:41]: Oh nice. I didn't know. Will use `test_cmp`

File: t/t3025-resource-find.t [hunk 14]
```diff
@@ -0,0 +1,256 @@
+#!/bin/sh
+
+test_description='Test Resource Find On Tiny Machine Configuration'
+
+. $(dirname $0)/sharness.sh
+
+sock1_job="${SHARNESS_TEST_SRCDIR}/data/resource/jobspecs/find/test001.yaml"
+c1g1m1_job="${SHARNESS_TEST_SRCDIR}/data/resource/jobspecs/find/test002.yaml"
+node1_job="${SHARNESS_TEST_SRCDIR}/data/resource/jobspecs/find/test003.yaml"
+full_job="${SHARNESS_TEST_SRCDIR}/data/resource/jobspecs/find/test004.yaml"
```
[the remark this answers is not in the document]
Comment (dongahn) [reply t/t3025-resource-find.t:10]: No objection. I was just following the convention so far. I do agree a descriptive name is better.

File: resource/traversers/dfu_impl.cpp [hunk 15]
```diff
@@ -575,6 +575,71 @@ int dfu_impl_t::dom_dfv (const jobmeta_t &meta, vtx_t u,
...
+    int rc = -1;
+    int nchildren = 0;
+    f_out_edg_iterator_t ei, ei_end;
+    expr_eval_vtx_target_t vtx_target;
+    std::string dom = m_match->dom_subsystem ();
+    bool result = false;
+    bool down = (*m_graph)[u].status == resource_pool_t::status_t::DOWN;
+    bool allocated = !(*m_graph)[u].schedule.allocations.empty ();
+    bool reserved = !(*m_graph)[u].schedule.reservations.empty ();
+    Flux::resource_model::vtx_predicates_override_t p_overriden = p;
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/traversers/dfu_impl.cpp:599]: Yes. `p_overriden` is first copy-constructed from `p` and then updated as passed with the following line `set()`. This way each method frame will have its own copy of `p_overriden` without having to leak it to the caller frame (since p is passed by reference.). I may be missing something though.

File: resource/evaluators/expr_eval_vtx_target.cpp [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+        goto done;
+    }
+    if (p == "status")
+        rc = (x == "down" || x == "DOWN" || x == "up" || x == "UP")? 0 : -1;
+    else if (p == "sched-now")
+        rc = (x == "allocated"
+             || x == "ALLOCATED" || x == "free" || x == "FREE")? 0 : -1;
+    else if (p == "sched-future")
+        rc = (x == "reserved"
+             || x == "RESERVED" || x == "free" || x == "FREE")? 0 : -1;
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/evaluators/expr_eval_vtx_target.cpp:55]: Heh. Yeah I thought about it and got lazy. C++ standard library must have an easy to do this. Will add that support.

File: resource/evaluators/expr_eval_api.cpp [hunk 17]
```diff
@@ -0,0 +1,352 @@
...
+int expr_eval_api_t::validate_leaf (const std::string &e,
+                                    const expr_eval_target_base_t &target)
+{
+    int rc = -1;
+    std::size_t delim;
+    std::string key, value;
+
+    m_level++;
+
+    if ( (delim = e.find_first_of ("=")) == std::string::npos) {
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/evaluators/expr_eval_api.cpp:141]: The intent was to say `p=x` key-value pair encodes `p(x)` in predicate calculus. We probably don't want a predicate actually to be expressed as p(x) since the double use of parenthesis, using `p=x` to encode p(x) would be okay? or is it too confusing?

File: resource/evaluators/expr_eval_api.cpp [hunk 18]
```diff
@@ -0,0 +1,352 @@
...
+}
+
+
+/****************************************************************************
+ *                                                                          *
+ *           Expression Evaluation API Public Method Definitions            *
+ *                                                                          *
+ ****************************************************************************/
+
+int expr_eval_api_t::validate (const std::string &e,
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/evaluators/expr_eval_api.cpp:279]: Yeah, this is possible and good idea. I think we can add `expr_eval_err_t` in a similar way to jansson (was it `json_err_t`?) and support you are asking for . For the interest of time, though, we may create a ticket for this and address it as part of a future small enhancement ticket?

File: resource/evaluators/expr_eval_api.hpp [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+
+    /* Evaluate methods */
+    int evaluate_leaf (const std::string &expr,
+                       const expr_eval_target_base_t &target, bool &result);
+    int evaluate_paren (const std::string &expr,
+                        const expr_eval_target_base_t &target,
+                        size_t at, size_t &next, bool &result);
+    int evaluate_pred (pred_op_t op, bool result2, bool &result1) const;
+
+    int m_level = 0;
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/evaluators/expr_eval_api.hpp:102]: The intent was to add some nice indented debug message using that. But yes since we don't use this in this PR, I will remove the use of this member variable.

File: resource/evaluators/expr_eval_api.cpp [hunk 20]
```diff
@@ -0,0 +1,352 @@
...
+    if ( (nws = e.find_first_not_of (" \t", at)) == std::string::npos) {
+        errno = EINVAL;
+        goto done;
+    }
+    if ( (xws = e.find_first_of (" \t", nws)) == std::string::npos)
+        xws = e.length ();
+    op = pred_op_t::AND;
+    next = at;
+    if (e.substr (nws, xws - nws) == "and") {
+        next = xws;
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/evaluators/expr_eval_api.cpp:122]: I haven't test it but `X=foo pr Y=bar` should be interpreted as `X=foo and pr and Y=bar` and by the time you get to `pr`, the evaluation system will invalidate the expression. I will test.

[the remark this answers is not in the document]
Comment (dongahn) [reply resource/evaluators/expr_eval_api.cpp:122]: Just confirmed it is the case:

```console
./resource-query -L../../t/data/resource/grugs/tiny.graphml
INFO: Loading a matcher: CA
resource-query> find status=up pr sched-now=allocated
ERROR: find: invalid criteria: status=up pr sched-now=allocated.
```

File: resource/evaluators/test/expr_eval_test01.cpp [hunk 21]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+int expr_eval_test_target_t::evaluate (const std::string &p,
+                                       const std::string &x, bool &result) const
+{
+    int rc = -1;
+    if ( (rc = validate (p, x)) < 0)
+        return rc;
+    result = true;
+    return rc;
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/evaluators/test/expr_eval_test01.cpp:66]: Hmm. The intent is for this function always to return the same return code as `validate`. And when it returns 0, `result` is always set to be true for testing purpose.

File: resource/evaluators/test/expr_eval_test01.cpp [hunk 22]
```diff
@@ -0,0 +1,205 @@
...
+        "status=up sched-now=allocated (( sched=free status=down)");
+    expr_vector.push_back ("status=up and sched-now=foo");
+    expr_vector.push_back ("sched-now=free status=down and");
+    expr_vector.push_back ("sched-now=free status=d own");
+    expr_vector.push_back ("(status=up (sched-now=allocated) "
+        "and (sched-now=free and) status=down)");
+    expr_vector.push_back ("(status=up (sched-now=allocated) "
+        "and (sched-now=free)status=down)");
+    expr_vector.push_back ("(status=up)(sched-now=allocated)"
+        "(sched-now=free)(status=down)");
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/evaluators/test/expr_eval_test01.cpp:130]: Yes, you need either "and" or "or" or one or more whitespaces which is implicit "and". -- Modeled after Unix `find` command.

File: resource/traversers/dfu_impl.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/traversers/dfu_impl.cpp:599]: I will mark it resolved for now. Let me know if I missed something.

File: resource/evaluators/expr_eval_api.cpp (hunk 12 above)
Comment (dongahn) [reply resource/evaluators/expr_eval_api.cpp:292]: I just realized this. I will add some to our unit tests.

File: resource/evaluators/expr_eval_vtx_target.cpp [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+int expr_eval_vtx_target_t::evaluate (const std::string &p,
+                                      const std::string &x, bool &result) const
+{
+    int rc = 0;
+
+    if ( (rc = validate (p, x)) < 0)
+        goto done;
+    if (p == "status") {
+        if (x == "down" || x == "DOWN") {
+            result = m_overriden.status_down
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/evaluators/expr_eval_vtx_target.cpp:71]: Great catch!

File: resource/evaluators/expr_eval_vtx_target.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/evaluators/expr_eval_vtx_target.cpp:55]: `std::transform` does the magic.

```diff
@@ -41,18 +43,19 @@ int expr_eval_vtx_target_t::validate (const std::string &p,
                                       const std::string &x) const
 {
     int rc = -1;
+    std::string lcx = x;
+
     if (!m_initialized) {
         errno = EINVAL;
         goto done;
     }
+    std::transform (x.begin(), x.end(), lcx.begin(), ::tolower);
     if (p == "status")
-        rc = (x == "down" || x == "DOWN" || x == "up" || x == "UP")? 0 : -1;
+        rc = (lcx == "down" || lcx == "up")? 0 : -1;
     else if (p == "sched-now")
```

File: resource/evaluators/expr_eval_api.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/evaluators/expr_eval_api.cpp:273]: Open Issue #699.

**Comment by dongahn** [comment]
@SteVwonder: I believe I have addressed all the issues. Could you take a one last look at the fixup commits (since this is pretty late at night I'm not that capable of looking at those with confidence :-)

Also I opened Issue at https://github.com/flux-framework/flux-sched/issues/699 per your request at https://github.com/flux-framework/flux-sched/pull/693#discussion_r457792401

File: resource/evaluators/expr_eval_api.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (SteVwonder) [reply resource/evaluators/expr_eval_api.cpp:139]: Ah! I think the thing I was missing was that that is how it is represented in predicate calculus.  I totally agree we don't want the extra parenthesis, lest this become a lisp variant ;)

File: resource/traversers/dfu_impl.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (SteVwonder) [reply resource/traversers/dfu_impl.cpp:599]: It might be worth throwing a comment above this line calling out the copy-construction for those less familiar with C++ (like me).  I thought this was just a simple name change, I didn't realize there was copying going on.  Thanks for the clarification.

**Comment by SteVwonder** [comment]
Looks like the valgrind failure was just intermittent.  All good for a squash and merge :)

File: resource/evaluators/expr_eval_api.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (milroy) [reply resource/evaluators/expr_eval_api.cpp:1]: What I meant to ask was whether this PR evaluates nested parenthetic expressions via recursion (which certainly didn't come across in my comment).  That's why I listed the call sequence.  This is a more of a curiosity and not necessary for the merge.

[the remark this answers is not in the document]
Comment (dongahn) [reply resource/evaluators/expr_eval_api.cpp:1]: Ah... sorry about that. Yes, it is handled through recursion.

**Comment by dongahn** [comment]
Thanks @SteVwonder and @milroy. I squashed all fixup commits and forced a push. When CI turns green, this will go in.
